capstone-frontend/src/components/SearchUserModal.jsx

```diff
@@ -0,0 +1,255 @@
+import React, { Component } from 'react';
+
+import { Button, Form, Spinner, Modal, Col, Row, Card } from 'react-bootstrap'
+
+import '../styles/SearchUserModal.css'
+import '../styles/Modal.css'
+
+export class SearchUserModal extends Component {
+
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      fetchingUsers: false, // For Spinner
+      showModal: false,
+      typingTimeout: 0,
+      searchTerm: '',
+      searchResults: [],
+      addedUsersTracker: [],
+      addedUsers: [],
+      addedFriends: [],
+      resultsFound: false,
+    }
+  }
+
+  getBooks = async (searchTerm) => {
+
+    this.setState({ fetchingUsers: true })
+
+    let searchResults;
+
+    if (searchTerm === '') {
+      searchResults = [];
+
+      this.setState({ searchResults, fetchingUsers: false, resultsFound: false })
+    }
+    else {
+      searchResults = await fetch(`/api/userSearch?searchTerm=${searchTerm}`)
+        .then(response => response.json())
+        .catch(err => alert(err));
+
+      if (typeof searchResults === 'undefined') {
+        searchResults = [];
+      }
+
+      if (searchResults.length !== 0) {
+        this.setState({ searchResults, fetchingUsers: false, resultsFound: true }) // Guilty
+      }
+      else {
+        this.setState({ searchResults, fetchingUsers: false, resultsFound: false }) // Not Guilty
+      }
+    }
+  }
+
+  handleSearchTermChange = (event) => {
+
+    // resultsFound innocent until proven Guilty
+    this.setState({ searchResults: [], resultsFound: true })
+
+    if (this.state.typingTimeout) {
+      clearTimeout(this.state.typingTimeout);
+    }
+
+    this.setState({
+      searchTerm: event.target.value,
+      typingTimeout: setTimeout(async () => {
+        await this.getBooks(this.state.searchTerm)
+      }, 500)
+    })
+  }
+
+  addUserToAddedUsers = (user) => {
+
+    if (!this.state.addedUsersTracker.includes(user)) {
+      this.setState({ addedUsersTracker: [...this.state.addedUsersTracker, user] })
+    }
+
+    // Rerender
+    this.setState({ addedUsers: [...this.state.addedUsers, user] })
+  }
+
+  addUserToAddedFriends = (user) => {
+    if (!this.state.addedUsersTracker.includes(user)) {
+      this.setState({ addedUsersTracker: [...this.state.addedUsersTracker, user] })
+    }
+
+    // Rerender
+    this.setState({ addedFriends: [...this.state.addedFriends, user] })
+  }
```

Comment:
What's the purpose of the `addedUsersTracker` in these methods? It looks like you could omit it and set `addedUsers` and `addedFriends` directly in the `if` cases.

Reply:
```addedUsersTracker``` is a helper array I use to render the ```addedUsers``` or ```addedFriends``` arrays. Without it, I would have to render both arrays separately, which may render duplicate data. In these functions, ```addedUsersTracker``` is checking if the it contains the user being added. These check stop ```addedUsersTracker``` from adding the same user twice.